### PR TITLE
feat(renderer): allow placing git icons after filename

### PR DESCRIPTION
**Pull request by kyazdani42**
This is a first draft, i'd like to improve the code a bit before merging because it's not pretty as it is, but usable.
This also change the behavior for special icons, which will display icons normally instead of the default file icon.

TODO:
- [x] config option documentation
- [x] improve the algorithms for placing the git icon
- [x] add the git icon for the symlinks also

implements https://github.com/kyazdani42/nvim-tree.lua/issues/246, https://github.com/kyazdani42/nvim-tree.lua/issues/1069 and https://github.com/kyazdani42/nvim-tree.lua/issues/72, #1217

### CODE REVIEW

File: lua/nvim-tree/renderer/builder.lua [hunk 1]
```diff
@@ -170,15 +182,24 @@ end
 function Builder:_build_file(node, padding, git_highlight)
   local offset = string.len(padding)
 
-  local icon, git_icons = self:_build_file_icons(node, offset)
+  local icon = self:_build_file_icon(node, offset)
```
Comment (kyazdani42) [inline lua/nvim-tree/renderer/builder.lua:185]: offset should have + #git_icons if git icons are placed before the filename.

**Comment by gegoune** [comment]
Screenshot 2022-04-27 at 00 55 33
Will check if master is the same but seeing some odd highlighting on that PR.

**Comment by kyazdani42** [comment]
i guess, see my comment before :) this is not perfect yet, the algorithms are very clunky as it is.

**Comment by gegoune** [comment]
Wasn't complaining, just reporting, PR is in progress, of course. :)

**Comment by kyazdani42** [comment]
@gegoune could you give it another try ? i've fixed multiple issues and also added icons for symlinks :)

File: lua/nvim-tree/renderer/builder.lua [hunk 2]
```diff
@@ -118,76 +126,96 @@ function Builder:_build_folder(node, padding, git_hl)
...
   self:_insert_line(line)
 end
 
-function Builder:_build_file_icons(node, offset)
+function Builder:_build_file_icon(node, offset)
   if self.special_map[node.absolute_path] or self.special_map[node.name] then
-    local git_icons = self:_unwrap_git_data(git.get_icons(node), offset + #icons.i.special)
-    self:_insert_highlight("NvimTreeSpecialFile", offset + #git_icons)
-    return icons.i.special, git_icons
+    self:_insert_highlight("NvimTreeSpecialFile", offset, offset + #icons.i.special + #node.name)
```
Comment (kyazdani42) [inline lua/nvim-tree/renderer/builder.lua:165]: this `+ #node.name` seems incorrect

**Comment by gegoune** [comment]
> @gegoune could you give it another try ? i've fixed multiple issues and also added icons for symlinks :)

Highlighting issue seems to be fixed now.

Unrelated, and I think there already is an issue for it, but why not put those markers in signcolumn?

**Comment by alex-courtis** [comment]
Nice. I'll give this a test / review on the weekend.

File: lua/nvim-tree/renderer/components/icons.lua [hunk 3]
```diff
@@ -94,7 +86,6 @@ function M.reset_config(webdev_colors)
   M.webdev_colors = webdev_colors
 
   config_symlinks()
-  config_special_icon()
```
Comment (alex-courtis) [inline lua/nvim-tree/renderer/components/icons.lua:97]: Removing this is good.

File: lua/nvim-tree/renderer/builder.lua [hunk 4]
```diff
@@ -118,76 +126,93 @@ function Builder:_build_folder(node, padding, git_hl)
...
-  self:_insert_highlight(foldername_hl, offset + #icon + #git_icon, #line)
+  self:_insert_highlight(foldername_hl, fname_starts_at, fname_starts_at + #foldername)
 
   if git_hl then
-    self:_insert_highlight(git_hl, offset + #icon + #git_icon, #line)
+    self:_insert_highlight(git_hl, fname_starts_at, fname_starts_at + #foldername)
   end
 end
 
--- TODO: missing git icon for symlinks
```
Comment (alex-courtis) [inline lua/nvim-tree/renderer/builder.lua:128]: buildAdd this comment back. `git_icons_tbl` is always empty for symlinks.

I have tested that git icons for symlinks are working for `_build_symlink` by manually injecting them e.g.
```lua
  git_icons_tbl = {
    {
      hl = "NvimTreeGitDirty",
      icon = "✗"
    }
  }
```

File: README.md [hunk 5]
```diff
@@ -153,7 +153,8 @@ require'nvim-tree'.setup { -- BEGIN_DEFAULT_OPTS
     },
     icons = {
       webdev_colors = true,
-    },
+      git_placement = "before"
```
Comment (alex-courtis) [inline README.md:156]: I see `git_placement = "sign_column"` in the future.

**Review by alex-courtis — APPROVED** [review]
This definitely works. Tested files and folders: before, after, foo

**Comment by alex-courtis** [comment]
RE symlink git icons:

They are not really part of this change. They can be made to work with some tweaking of `node-builders.link`, conditionally using the absolute or link path and `get_dir_git_status` for folder links.

Raised #1217

File: lua/nvim-tree/renderer/builder.lua (hunk 4 above)
Comment (kyazdani42) [reply lua/nvim-tree/renderer/builder.lua:128]: if you add a symlink into a git directory it will not be empty :)

File: README.md (hunk 5 above)
Comment (kyazdani42) [reply README.md:156]: :+1:

**Comment by kyazdani42** [comment]
@alex-courtis do you want me to remove the code that adds the git icons for symlinks ? It works fine on my end though.

**Comment by alex-courtis** [comment]
> @alex-courtis do you want me to remove the code that adds the git icons for symlinks ? It works fine on my end though.

No! I've tested it. Once we resolve #1217 everything will display correctly.

File: lua/nvim-tree/renderer/builder.lua (hunk 4 above)
Comment (alex-courtis) [reply lua/nvim-tree/renderer/builder.lua:128]: maybe add a comment
```
-- TODO: missing git icon for symlinks #1217

**Comment by alex-courtis** [comment]
Test harness `nvim -nu /tmp/nvt-min.lua`
[nvt-min.lua.txt](https://github.com/kyazdani42/nvim-tree.lua/files/8600650/nvt-min.lua.txt)

Install [ccls](https://github.com/MaskRay/ccls). `pacman -S ccls`. Not sure what package manager you use; needs clang and node.

Clone https://github.com/alex-courtis/way-displays

File: lua/nvim-tree/renderer/builder.lua (hunk 4 above)
Comment (kyazdani42) [reply lua/nvim-tree/renderer/builder.lua:128]: git icons are well present for symlinks.

**Pull request merged**